### DISCUSSION TITLE
privatelink

### DISCUSSION THREAD

**Discussion by franwitter** [Snowflake-Labs/terraform-provider-snowflake#370] [Q&A]
Does the snowflake provider work with privatelinks?

**Comment by franwitter** [comment]
I figured out a work around.  I concatenated ".privatelink" to region and it works.

**Comment by dsalaj** [comment]
To get available values for privatelink for your account you can use the following query:
```
use role accountadmin;
select key, value from table(flatten(input=>parse_json(system$get_privatelink_config())));
```
for further details if you need exact host and port use this query:
```
SELECT t.VALUE:type::VARCHAR as type,
       t.VALUE:host::VARCHAR as host,
       t.VALUE:port as port
FROM TABLE(FLATTEN(input => PARSE_JSON(SYSTEM$ALLOWLIST_PRIVATELINK()))) AS t;
```

And finally beware that the privatelinks use large IP Ranges, in addition to RFC1918, also the future use range:
```
  '10.0.0.0/8'
  '172.16.0.0/12'
  '192.168.0.0/16'
  '240.0.0.0/4'
```

To check if your configured service users are actually using the privatelink connection, check the `LOGIN_HISTORY` using this query:
```
select max(EVENT_TIMESTAMP) as last_login, CLIENT_IP, user_name, count(*) as number_of_logins
from SNOWFLAKE.ACCOUNT_USAGE.LOGIN_HISTORY
where user_name = 'YOUR_SERVICE_USER_NAME'
group by CLIENT_IP, user_name
order by last_login desc;
```
